src/promptflow/promptflow/executor/_async_nodes_scheduler.py

```diff
@@ -137,6 +156,62 @@ def signal_handler(sig, frame):
     raise KeyboardInterrupt
 
 
+def log_stack_recursively(task: asyncio.Task, elapse_time: float):
+    """Recursively log the frame of a task or coroutine.
+    Traditional stacktrace would stop at the first awaited nested inside the coroutine.
+
+    :param task: Task to log
+    :type task_or_coroutine: asyncio.Task
+    :param elapse_time: Seconds elapsed since the task started
+    :type elapse_time: float
+    """
+    messages = [f"Task {task.get_name()} has been running for {elapse_time:.0f} seconds, stacktrace:"]
+    task_or_coroutine = task
+    while True:
+        if isinstance(task_or_coroutine, asyncio.Task):
+            # For a task, get the coroutine it's running
+            coroutine: asyncio.coroutine = task_or_coroutine.get_coro()
+        elif asyncio.iscoroutine(task_or_coroutine):
+            coroutine = task_or_coroutine
+        else:
+            flow_logger.info("\n".join(messages))
+            return
+
+        frame = coroutine.cr_frame
+        if frame is not None:
+            messages.append(str(inspect.getframeinfo(frame)))
+
+        # Follow to the next awaited coroutine, if any
+        task_or_coroutine = coroutine.cr_await
+
+
+def monitor_long_running_coroutine(loop: asyncio.AbstractEventLoop, task_start_time: dict, task_last_log_time: dict):
+    flow_logger.info("monitor_long_running_coroutine started")
+    while True:
+        running_tasks = [task for task in asyncio.all_tasks(loop) if not task.done()]
```

Comment:
No break here?

Reply:
The daemon thread will exit with the main process, there seems to be no need to break.
And the exit condition would be tricky, since no running_tasks may not indicate the flow is done.

Reply:
Changed to use environment variable to let main thread tell mointor thread to stop when the dag is completed.